### PR TITLE
[WIP] add new segmentation algorithm

### PR DESCRIPTION
**Pull request by yuyu2172**
# What I have done

I implemented a segmentation algorithm that is described here. [paper](http://www.robotics.tu-berlin.de/fileadmin/fg170/Publikationen_pdf/Jonschkowski-16-Technical-Report.pdf)

Some part of the algorithm and a public data set are shared [here](https://gitlab.tubit.tu-berlin.de/rbo-lab/rbo-apc-object-segmentation).

I implemented data pre-processing codes and ROS interface which lacked in the public code. 

The algorithm takes RGBD information as input and produce segmented image as output. The first picture illustrates output. The second image illustrates internal data that evaluates probability of pixel assigned to each class. The figures in the second image are rotaed 90 degrees. 

2016_04_24_20_12_bin_g_cheezit_big_original

screenshot from 2016-04-25 10 55 02

# TODOs
+ The images are taken by kinect2 on torso. Run algorithm on data taken from softkinetic camera.
+ Take more examples and analyze performance.

### CODE REVIEW

**Comment by wkentaro** [comment]
this can be installed in jsk_apc2016_common package, cant it? because not robot specific.
and you can install python modules as python package using setup.py with catkin_setup_python cmake function.

**Comment by yuyu2172** [comment]
@wkentaro 
Yes, this is not specific to robot. I will try your comment.

**Comment by k-okada** [comment]
i think we'd better to put this into jsk_recognition,

**Comment by wkentaro** [comment]
+1 to @k-okada's suggestion, if you have time to split the code to APC specific and general image processing.

**Comment by yuyu2172** [comment]
@k-okada , @k-okada 
I agree that the code should be put in jsk_recognition.
However, I think that relocating the code should not be done until it stabilizes.
For the time being, I would like to relocate the code to jsk_apc2016_common. Besides, most of the code is specific to APC settings (ex: feature extraction that uses model of APC shelf).

@k-okada 
Can members of APC2016 have permission to merge pull requests?

**Comment by k-okada** [comment]
jsk_recognitoin is more ore less unstable codes, if you can write some
codes that with confident, please move to
http://wiki.ros.org/opencv_apps?distro=jade

--
◉ Kei Okada

**Comment by k-okada** [comment]
I think you first try to send pull request to https://gitlab.tubit.tu-berlin.de/rbo-lab/rbo-apc-object-segmentation and put link to that PR pages

File: .travis [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit ef20f1b17aef8f0e2225cf52689f6fa5f0bfb504
+Subproject commit 7bd3fa251a90d50e7ee15af0c767974746bb4243
```
Comment (knorth55) [inline .travis]: `git submodule update`

File: jsk_2016_01_baxter_apc/package.xml [hunk 2]
```diff
@@ -36,6 +37,5 @@
   <run_depend>roscpp</run_depend>
   <run_depend>std_msgs</run_depend>
 
-  <test_depend>rostest</test_depend>
```
Comment (knorth55) [inline jsk_2016_01_baxter_apc/package.xml]: you do not need to modify this

File: jsk_2016_01_baxter_apc/node_scripts/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
+#!/usr/bin/env python
+# TODO: speed up spatial
```
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/pipeline.py]: file name is too general, more specified name is preferred

File: jsk_2016_01_baxter_apc/node_scripts/helper.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
```
Comment (knorth55) [inline jsk_2016_01_baxter_apc/node_scripts/helper.py]: too general name, more specified name is preferred

Comment (wkentaro) [inline jsk_2016_01_baxter_apc/node_scripts/helper.py]: This is python module and not a node script, right?
You can put this as python module with `catkin_setup_python()` in cmake and setup.py file.

**Comment by wkentaro** [comment]
> I think you first try to send pull request to https://gitlab.tubit.tu-berlin.de/rbo-lab/rbo-apc-object-segmentation and put link to that PR pages

@yuyu2172 and you can put the rbo code as submodule in this repo.

**Comment by yuyu2172** [comment]
@knorth55  @wkentaro 
Thanks for comments!

@k-okada 
I would like to know why you think I should do so. By the way, I am considering of contacting members of RBO by email if it is ok for us to use their code for the time being.

**Comment by yuyu2172** [comment]
Since this pull request was too big, I will divide it into pieces.

**Comment by k-okada** [comment]
> I would like to know why you think I should do so.
>
> - if we receive something, you'd better to give them back.

> By the way, I am considering of contacting members of RBO by email if it
is ok for us to use their code for the time being.

it is very nice for you to try communicate with authors, please do that.
But the license says you can use it ->
https://gitlab.tubit.tu-berlin.de/rbo-lab/rbo-apc-object-segmentation/blob/master/setup.py#L36

--
◉ Kei Okada

**Comment by yuyu2172** [comment]
@k-okada 
Thank you for the comment!

**Pull request closed without merging**